pennylane_lightning/experimental/lightning_qubit_2.py

```diff
@@ -0,0 +1,171 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the next generation successor to LightningQubit
+"""
+
+import numpy as np
+from typing import Union, Callable, Tuple, Optional, Sequence
+from warnings import warn
+
+from pennylane.devices.experimental import Device
+from pennylane.tape import QuantumTape, QuantumScript
+from pennylane.devices.experimental.execution_config import ExecutionConfig, DefaultExecutionConfig
+from pennylane.devices.qubit.simulate import simulate
+from pennylane.devices.qubit.preprocess import preprocess
+
+QuantumTapeBatch = Sequence[QuantumTape]
+QuantumTape_or_Batch = Union[QuantumTape, QuantumTapeBatch]
+
+try:
+    from ..lightning_qubit_ops import StateVectorC64, StateVectorC128
+
+    CPP_BINARY_AVAILABLE = True
+except ModuleNotFoundError:
+    CPP_BINARY_AVAILABLE = False
+
+if CPP_BINARY_AVAILABLE:
+    DeviceExecutionConfig = DefaultExecutionConfig
+
+    class LightningQubit2(Device):
+        """PennyLane Lightning device.
+
+        A device that interfaces with C++ to perform fast linear algebra calculations.
+
+        Use of this device requires pre-built binaries or compilation from source. Check out the
+        :doc:`/installation` guide for more details.
+
+        Args:
+            c_dtype: Datatypes for statevector representation. Must be one of ``np.complex64`` or ``np.complex128``.
+        """
+
+        def __init__(self, c_dtype=np.complex128):
```

Comment:
While we don't need to solve this now, I think having the option of a stateful memory block as part of the device as a config option to be fed in from elsewhere, could be useful. I still prefer the allocate-and-run strategy, but that may not always be preferential. Let's come back to it another time.